### PR TITLE
DOC: Added 'Modules Privacy Has Changed' section to overview.rst

### LINKED ISSUE #22942: Document module privacy
We should better document what in pandas is considered private.

http://pandas.pydata.org/pandas-docs/stable/whatsnew.html#reorganization-of-the-library-privacy-changes should be captured somewhere more permanent than in a whatsnew release. source is at https://github.com/pandas-dev/pandas/blob/master/doc/source/whatsnew/v0.20.0.txt#L1194

I think a new section in http://pandas-docs.github.io/pandas-docs-travis/overview.html is the best place for it.

### PR DESCRIPTION
**Pull request by aparimeet**
- [x] closes #22942 
- [x] tests passed

### CODE REVIEW

File: doc/source/overview.rst [hunk 1]
```diff
@@ -117,6 +117,46 @@ Institutional Partners
 
 The information about current institutional partners can be found on `pandas website page <https://pandas.pydata.org/about.html>`__.
 
+Modules Privacy Has Changed
```
Comment (TomAugspurger) [inline doc/source/overview.rst]: Remove the "Has Changed"

**Review by TomAugspurger — COMMENTED** [review]
High-level feedback: could you remove all references to "have been moved" or "has changed". The release note you based this off is meant for readers upgrading between versions of pandas. We want this to be a timeless document that describes things in the current version.

So instead of things like "... modules have been moved"  we really care about the parts like

```
The following modules in pandas are considered private:

* pandas.compat
* pandas.util
* pandas.core (and everything under it)
```

I don't think the table is necessary.

File: doc/source/overview.rst [hunk 2]
```diff
@@ -117,6 +117,46 @@ Institutional Partners
...
+----------------------------
+
+Some formerly public python/c/c++/cython extension modules have been moved and/or renamed. These are all removed from the public API. Furthermore, the pandas.core, pandas.compat, and pandas.util top-level modules are now considered to be PRIVATE. If indicated, a deprecation warning will be issued if you reference theses modules. (GH12588)
```
Comment (jschendel) [inline doc/source/overview.rst]: Can you replace (GH12588) with (:issue:\`12588\`) here, and likewise with the other instances of this pattern elsewhere.  I don't think the GH pattern will actually link to the issues.

File: doc/source/overview.rst (hunk 1 above)
Comment (aparimeet) [reply doc/source/overview.rst]: Sorry, didn't think about that. I'll get right on it.

File: doc/source/overview.rst (hunk 2 above)
Comment (aparimeet) [reply doc/source/overview.rst]: Alright, I'll fix that.

**Comment by aparimeet** [comment]
Let me know if this is correct.

File: doc/source/overview.rst [hunk 3]
```diff
@@ -117,6 +117,30 @@ Institutional Partners
 
 The information about current institutional partners can be found on `pandas website page <https://pandas.pydata.org/about.html>`__.
 
+Modules Privacy
+----------------
+
+The following modules in pandas are considered PRIVATE:
+
+::
```
Comment (TomAugspurger) [inline doc/source/overview.rst]: Does the double colon format this as a code block? If so, I don't think we need that. Sphinx should render the nested list fine.

File: doc/source/overview.rst [hunk 4]
```diff
@@ -117,6 +117,30 @@ Institutional Partners
 
 The information about current institutional partners can be found on `pandas website page <https://pandas.pydata.org/about.html>`__.
 
+Modules Privacy
+----------------
+
+The following modules in pandas are considered PRIVATE:
```
Comment (TomAugspurger) [inline doc/source/overview.rst]: Two things

1. A bit of context would be helpful
2. I worry about this going out of date

So how about something like

    In addition to sub-packages and modules prefixed with `_`, the following sub-packages and modules are considered private and may change or functionality without warning.
    
    * `pandas.compat`
    * `pandas.core` and all submodules
    * `pandas.util` and all submodules

    While pandas developers have always considered everything in `pandas.core` "private", we deemed simply renaming it to `pandas._core` too disruptive.

    Pandas' public API consists of everything listed in :ref:`api`.

Alternatively, we can list out modules

Comment (TomAugspurger) [reply doc/source/overview.rst]: Agreed.

File: doc/source/overview.rst (hunk 3 above)
Comment (TomAugspurger) [reply doc/source/overview.rst]: @aparimeet can you check how this renders without the double-colons? A regular nested list should be fine.

You may need to remove the line-breaks between top-level items, and put spaces before the `-`.

**Review by TomAugspurger — APPROVED** [review]
Updated the modules.

I think we just need to list `pandas.core` and `pandas.compat`.

Thisn like `pandas.util.decorators` are deprecated, so there's no question whether they're in the public API.

**Review by jorisvandenbossche — COMMENTED** [review]
I think the text around it suggested by Tom in https://github.com/pandas-dev/pandas/pull/23038#discussion_r223694209 would still be good to add

Do we have a better place for this than `overview.rst` ? It now feels a bit out of place between "Institutional partners" and "License"

**Comment by TomAugspurger** [comment]
The neighboring sections are a bit strange, but "Package Overview" seems
like the right place.

Maybe we could expand the new section a bit with text describing the
modules that are public?

**Comment by aparimeet** [comment]
Adding public modules to this would be a good idea.

**Comment by datapythonista** [comment]
I'm unsure about this. I see the value in documenting what is private or public, but:
- I agree that between partners and license doesn't seem to be the right place
- `api.rst` is already the list of what is public, and it's huge, we probably don't want to repeat this here, and I find quite confusing to have a public modules list with a subset

@TomAugspurger  What about moving this as a short `.. note::` section at the beginning of `api.rst` to warn about what is private instead?

**Comment by TomAugspurger** [comment]
I notice now that we *do* have a note like that at the top of api.rst:
http://pandas-docs.github.io/pandas-docs-travis/api.html

**Comment by datapythonista** [comment]
That's funny, didn't see that. Closing this then, feel free to reopen if you think something else is needed.

**Comment by TomAugspurger** [comment]
Sorry for missing that @aparimeet. Let me know if you want help starting on
a new issue.

**Pull request closed without merging**